File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        res = [] 

        def bst(root):
            if root.left:
                bst(root.left)
            res.append(root.val)
            if root.right:
                bst(root.right)
                
        bst(root)
        return res[k - 1] 

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    isSameTree(p, q) {
        if (!p && !q) {
            return true
        }

        if (!p || !q) {
            return false
        }

        if (p && q && p.val === q.val) {
            return this.isSameTree(p.left, q.left) && this.isSameTree(p.right, q.right)
        }

        return false
    }


    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if (!root && !subRoot) {
            return true
        }

        if (!root) {
            return false
        }

        if (this.isSameTree(root, subRoot)) {
            return true
        }

        return this.isSubtree(root.left, subRoot) || this.isSubtree(root.right, subRoot)
    }
}
